### PR TITLE
[BugFix] Avoid repeatedly adding tablet to pending queue

### LINKED ISSUE #7871: tablet scheduling may repeatedly add the same tablet to pending list
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by nshangyiming**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7871

## Problem Summary(Required) ：

        // Under no circumstance should we repeatedly add a tablet to pending queue
        // to schedule, because this will break the scheduling logic. Besides, with current design,
        // we have to maintain the constraint that `allTabletIds = runningTablets + pendingTablets`.
        // `allTabletIds` and `runningTablets` are defined as unique container, if we schedule a
        // tablet with different `TabletSchedCtx` at the same time, the reference in `runningTablets`
        // can be messed up.
        // `force` here should only mean that we can exceed the size limit of
        // `pendingTablets` and `runningTablets` if there is too many tablets to schedule.

we check `pendingList` before add a new tablet to it.

### CODE REVIEW

**Comment by nshangyiming** [comment]
run_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java [hunk 1]
```diff
@@ -393,6 +410,7 @@ private synchronized void adjustPriorities() {
     private void schedulePendingTablets() {
         long start = System.currentTimeMillis();
         List<TabletSchedCtx> currentBatch = getNextTabletCtxBatch();
+        debugLogPendingTabletsStats();
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java:413]: should check if LOG.isDebugLevel first before call this function?

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java [hunk 2]
```diff
@@ -233,10 +233,17 @@ public Map<Long, PathSlot> getBackendsWorkingSlots() {
...
-        if (!force && containsTablet(tablet.getTabletId())) {
+        // Under no circumstance should we repeatedly add a tablet to pending queue
+        // to schedule, because this will break the scheduling logic. Besides, with current design,
+        // we have to maintain the constraint that `allTabletIds = runningTablets + pendingTablets`.
+        // `allTabletIds` and `runningTablets` are defined as unique container, if we schedule a
+        // tablet with different `TabletSchedCtx` at the same time, the reference in `runningTablets`
+        // can be messed up.
+        // `force` here should only mean that we can exceed the size limit of
+        // `pendingTablets` and `runningTablets` if there is too many tablets to schedule.
+        if (pendingTablets.stream().filter(t -> t.getTabletId() == tablet.getTabletId()).count() != 0) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java:246]: a O(n) check will it be a bottleneck?

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java (hunk 1 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java:413]: yes, should check here

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java (hunk 2 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java:246]: size of `pendingTablets`  is limited(maximum is 2000 by default) and we already have 4 levels of nested loop in `TabletChecker`..., so i think it's ok but need further test.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 14 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|0|13|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|0|1|00.00%|

**Review by imay — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java (hunk 2 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java]: i think we can avoid O(n) search in `pendingTablets` and still make use of `allTabletIds` map, although in a specific situation we should explicitly remove the tablet id from `allTabletIds` first before `addTablet()`, see the function `addBackToPendingTablets()` for details.

**Comment by nshangyiming** [comment]
run_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 18 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|0|17|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|0|1|00.00%|

**Review by imay — DISMISSED** [review]

**Comment by kangkaisen** [comment]
How did you test this bug and where are the test cases?

**Review by chaoyli — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java [hunk 3]
```diff
@@ -407,6 +427,8 @@ private void schedulePendingTablets() {
                 tabletCtx.setErrMsg(e.getMessage());
 
                 if (e.getStatus() == Status.SCHEDULE_FAILED) {
+                    LOG.info("scheduling for tablet[{}] failed, type: {}, reason: {}",
+                            tabletCtx.getTabletId(), tabletCtx.getType().name(), e.getMessage());
```
[the remark this answers is not in the document]
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletScheduler.java:431]: May print too many logs when BE path slot is busy, i will change it to DEBUG level.

**Comment by nshangyiming** [comment]
> How did you test this bug and where are the test cases?

I tested in a test deployment, acutally it's more like a refinement than bugfix, because the code logic currently won't add two tablet with the same id into pending list. But the meaning of  `force` argument here is vague, we may accidentally do stupid things in the future, i think we should make some prevention.
Anyway, i shall look into our integration test repo and see how to add some tests there.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 18 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletScheduler.java|0|17|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|0|1|00.00%|

**Review by chaoyli — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Pull request merged**